### PR TITLE
Save memory when histogram agg is not on top

### PR DESCRIPTION
**Pull request by nik9000**
This saves some memory when the `histogram` aggregation is not a top
level aggregation by dropping `asMultiBucketAggregator` in favor of
natively implementing multi-bucket storage in the aggregator. For the
most part this just uses the `LongKeyedBucketOrds` that we built the
first time we did this.

### CODE REVIEW

**Comment by nik9000** [comment]
I'm going to add a test for the new debug information.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/AbstractHistogramAggregator.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Map;
+import java.util.function.BiConsumer;
+
+/**
+ * Base class for functionality shared between aggregators for this
+ * {@code histogram} aggregation.
+ */
+public abstract class AbstractHistogramAggregator extends BucketsAggregator {
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/AbstractHistogramAggregator.java:43]: There was a TODO that the range and numeric version of the aggregator shared a ton of code. Now they share a superclass that provides all that code.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/HistogramAggregatorFactory.java [hunk 2]
```diff
@@ -52,36 +51,11 @@
...
     static void registerAggregators(ValuesSourceRegistry.Builder builder) {
         builder.register(HistogramAggregationBuilder.NAME, CoreValuesSourceType.RANGE,
-            new HistogramAggregatorSupplier() {
-                @Override
-                public Aggregator build(String name, AggregatorFactories factories, double interval, double offset,
-                                        BucketOrder order, boolean keyed, long minDocCount, double minBound, double maxBound,
-                                        ValuesSource valuesSource, DocValueFormat formatter, SearchContext context,
-                                        Aggregator parent, Map<String, Object> metadata) throws IOException {
-                    ValuesSource.Range rangeValueSource = (ValuesSource.Range) valuesSource;
-                    if (rangeValueSource.rangeType().isNumeric() == false) {
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/HistogramAggregatorFactory.java:62]: I moved this check into the ctor so I can use the ctor reference that we've been doing elsewhere.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/HistogramAggregatorFactory.java [hunk 3]
```diff
@@ -117,10 +91,6 @@ protected Aggregator doCreateInternal(ValuesSource valuesSource,
                                             Aggregator parent,
                                             boolean collectsFromSingleBucket,
                                             Map<String, Object> metadata) throws IOException {
-        if (collectsFromSingleBucket == false) {
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/HistogramAggregatorFactory.java:120]: *this* is the important line!

File: server/src/test/java/org/elasticsearch/search/aggregations/bucket/histogram/NumericHistogramAggregatorTests.java [hunk 4]
```diff
@@ -55,11 +64,9 @@ public void testLongs() throws Exception {
             HistogramAggregationBuilder aggBuilder = new HistogramAggregationBuilder("my_agg")
                     .field("field")
                     .interval(5);
-            MappedFieldType fieldType = new NumberFieldMapper.NumberFieldType(NumberFieldMapper.NumberType.LONG);
-            fieldType.setName("field");
             try (IndexReader reader = w.getReader()) {
                 IndexSearcher searcher = new IndexSearcher(reader);
-                InternalHistogram histogram = search(searcher, new MatchAllDocsQuery(), aggBuilder, fieldType);
```
Comment (nik9000) [inline server/src/test/java/org/elasticsearch/search/aggregations/bucket/histogram/NumericHistogramAggregatorTests.java:62]: We do this kind of thing all over the place so I figured I'd make a utility method for it.

File: test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java [hunk 5]
```diff
@@ -864,4 +866,50 @@ private void cleanupReleasables() {
         Releasables.close(releasables);
         releasables.clear();
     }
+
```
Comment (talevy) [inline test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java:869]: nice touch. likely usable across many future tests

**Review by talevy — APPROVED** [review]

**Comment by nik9000** [comment]
I'm pulling some performance numbers for this. I'll likely merge before they get done and update with them once they come in. I'm fairly confident in it though.

**Pull request merged**